streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -159,7 +159,8 @@ public <K, V> void send(final String topic,
                     final Set<Integer> multicastPartitions = maybeMulticastPartitions.get();
                     if (multicastPartitions.isEmpty()) {
                         // If a record is not to be sent to any partition, mark it as a dropped record.
-                        log.debug("Not sending the record with key {} , value {} to any partition", key, value);
+                        log.warn("Skipping record as partitioner returned empty partitions. "
```

Comment:
Imo we shouldn't put anything that can potentially be logged for every record above `TRACE` as a rule of thumb. We always get user complaints and have to downgrade the logging level for these sort of things anyways. I would suggest to go back to including the key and value but put it at TRACE instead.

Thoughts? @cadonna

Reply:
@ableegoldman my two cents , i think it's safer to not log keys and values as those can contain sensitive information at times and can lead to security issues.

Reply:
I think, we should never log keys and values on any log level. Since the provider of the Streams service might not own the data, but might be able to change the log level, having keys and values in the logs seems a security risk to me.

Regarding, debug vs. warn vs. trace here. In other places, we use warn log messages when records are dropped since the dropped records metric does not contain any information about why the record was dropped. However, in this case I understand -- and correct me if I am wrong -- that the partitioner might drop records on purpose. So having a warn log message for each record seems not appropriate. Still it would be great to have some kind of logging. Maybe some regular summary (e.g., for the first 10 records and then every 100 records and then on larger intervals?) or indeed logging at trace level.

Reply:
My impression was that it's been standard practice to allow logging keys/values, but only at the TRACE level -- might be worth checking with the security team?

I guess my feeling here is that this definitely shouldn't be logged at WARN for every record, but could be done as a regular summary (a practice I think we should be adopting more often than we have in the past) or as I said, at TRACE. I would favor the summary unless we can include the key/value as a TRACE log, which again we should try to confirm the standard here. I suppose it's useful to know that records are being dropped due to the partitioner, which makes the WARN summary useful, but I do wonder if users might still have trouble debugging this partitioner choice without access to the key/values it's dropping?

Thoughts?